File: Tree/Leetcode-108-ConvertSortedArrayToBinarySearchTree/Leetcode-108-ConvertSortedArrayToBinarySearchTree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    TreeNode *sortedArrayToBST(vector<int> &nums)
    {
        if( nums.size() == 0 ) return nullptr;
        int mid = nums.size() / 2;
        TreeNode *root = new TreeNode(nums[mid]);
        vector<int> left, right;
        for (int i = 0; i < mid; i++) left.push_back(nums[i]);
        for (int i = mid + 1; i < nums.size(); i++) right.push_back(nums[i]);
        root->left = sortedArrayToBST(left);
        root->right = sortedArrayToBST(right);
        return root;
    }
};

File: maureen72/Tree/Leetcode-108-ConvertSortedArrayToBinarySearchTree.py
# Definition for a binary tree node.
class TreeNode(object):
     def __init__(self, val=0, left=None, right=None):
         self.val = val
         self.left = left
         self.right = right
class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        def constructBST(left, right):
            if left > right:
                return None
            
            mid = (left + right) // 2
            root = TreeNode(val=nums[mid])
            
            root.left = constructBST(left, mid - 1)
            root.right = constructBST(mid + 1, right)
            
            return root
        
        return constructBST(0, len(nums) - 1)